core/vm/evmc.go

```diff
@@ -0,0 +1,376 @@
+	origin = env.Origin
+	coinbase = env.Coinbase
+	number = env.BlockNumber.Int64()
+	timestamp = env.Time.Int64()
+	gasLimit = int64(env.GasLimit)
+	difficulty = common.BigToHash(env.Difficulty)
+
+	return gasPrice, origin, coinbase, number, timestamp, gasLimit, difficulty
+}
+
+func (host *HostContext) GetBlockHash(number int64) common.Hash {
+	env := host.env
+	b := env.BlockNumber.Int64()
+	if number >= (b-256) && number < b {
+		return env.GetHash(uint64(number))
+	}
+	return common.Hash{}
+}
+
+func (host *HostContext) EmitLog(addr common.Address, topics []common.Hash, data []byte) {
+	env := host.env
+	env.StateDB.AddLog(&types.Log{
+		Address:     addr,
+		Topics:      topics,
+		Data:        data,
+		BlockNumber: env.BlockNumber.Uint64(),
+	})
+}
+
+func (host *HostContext) Call(kind evmc.CallKind,
+	destination common.Address, sender common.Address, value *big.Int, input []byte, gas int64, depth int,
+	static bool, salt *big.Int) (output []byte, gasLeft int64, createAddr common.Address, err error) {
+
+	env := host.env
+
+	gasU := uint64(gas)
+	var gasLeftU uint64
+
+	switch kind {
+	case evmc.Call:
+		if static {
+			output, gasLeftU, err = env.StaticCall(host.contract, destination, input, gasU)
+		} else {
+			output, gasLeftU, err = env.Call(host.contract, destination, input, gasU, value)
+		}
+	case evmc.DelegateCall:
+		output, gasLeftU, err = env.DelegateCall(host.contract, destination, input, gasU)
+	case evmc.CallCode:
+		output, gasLeftU, err = env.CallCode(host.contract, destination, input, gasU, value)
+	case evmc.Create:
+		var createOutput []byte
+		createOutput, createAddr, gasLeftU, err = env.Create(host.contract, input, gasU, value)
+		isHomestead := env.ChainConfig().IsHomestead(env.BlockNumber)
+		if !isHomestead && err == ErrCodeStoreOutOfGas {
+			err = nil
+		}
+		if err == errExecutionReverted {
+			// Assign return buffer from REVERT.
+			// TODO: Bad API design: return data buffer and the code is returned in the same place. In worst case
+			//       the code is returned also when there is not enough funds to deploy the code.
+			output = createOutput
+		}
+	case evmc.Create2:
+		var createOutput []byte
+		createOutput, createAddr, gasLeftU, err = env.Create2(host.contract, input, gasU, value, salt)
+		if err == errExecutionReverted {
+			// Assign return buffer from REVERT.
+			// TODO: Bad API design: return data buffer and the code is returned in the same place. In worst case
+			//       the code is returned also when there is not enough funds to deploy the code.
+			output = createOutput
+		}
+	default:
+		panic(fmt.Errorf("EVMC: Unknown call kind %d", kind))
+	}
+
+	// Map errors.
+	if err == errExecutionReverted {
+		err = evmc.Revert
+	} else if err != nil {
+		err = evmc.Failure
+	}
+
+	gasLeft = int64(gasLeftU)
+	return output, gasLeft, createAddr, err
+}
+
+func getRevision(env *EVM) evmc.Revision {
+	n := env.BlockNumber
+	conf := env.ChainConfig()
+	if conf.IsPetersburg(n) {
```

Comment:
Please use a switch instead of so many ifs.

Reply:
I'm not sure how to do it. There is not a single value that can be the argument of a `switch()`.

Reply:
Go lets you have switches on more than one condition, you just leave the `switch` empty and write your conditions in the `case` statement

Reply:
Done.